File: 4 Kyu/4 kyu - range extraction.c
#include <stdlib.h>

#define BUFFSIZE 1024
#define MIN(x,y) (((x) < (y)) ? (x) : (y))
#define MAX(x,y) (((x) > (y)) ? (x) : (y))


char *range_extraction(const int *args, size_t n) {
    int min = args[n - 1], max = args[0], cnt = 0;
    char *output = malloc (BUFFSIZE * sizeof (char)), *out = output;
    *output = '\0';
    //printf ("size :::%i:::\n",n);
    for (int i = 0; i < n; ++i) {
        min = MIN (min, args[i]), max = MAX (max, args[i]);
        cnt++;
        if (args[i] + 1 != args[i + 1]) {

            if (cnt < 2)       out += sprintf (out,"%i",min);
            else if (cnt == 2) out += sprintf (out,"%i,%i",min,max);
            else               out += sprintf (out,"%i-%i",min,max);

            if (i + 1 < n)     out += sprintf (out,",");

            min = args[i + 1], cnt = 0;
        }
            //out += sprintf (out,"%i-%i,",min,max);
    }
    //printf ("%i ",args[i]);

    return output;
}

File: 4 Kyu/4 kyu - range extraction.rs
mod solution {
    pub fn range_extraction(args: &[i32]) -> String {

        let size = args.len();
        let mut minval:i32 = args[size - 1];
        let mut maxval:i32 = args[0];
        let mut cnt:i32 = 0;
        let mut nxt;
        let mut os = String::new();

        for i in 0..size {
            minval = std::cmp::min(minval, args[i]);
            maxval = std::cmp::max(maxval, args[i]);
            cnt += 1;
            if i < size - 1 { nxt = args[i+1] } else { nxt = 0 }

            if args[i] + 1 != nxt {

                if cnt < 2 {
                    os += &format! ("{minval}")
                } else if cnt == 2 {
                    os += &format! ("{minval},{maxval}")
                } else {
                    os += &format! ("{minval}-{maxval}")
                }

                if i + 1 < size { os += &format! (",") }

                minval = nxt;
                cnt = 0;
            }
        }

        os
    }
}
